### PR TITLE
Problem: memiavl background snapshot rewriting panic when shutdown

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- gracefully cancel the task when shutdown

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

## Summary by CodeRabbit

- **New Features**
  - Snapshot rewriting can now be gracefully canceled during shutdown.
- **Improvements**
  - Import operations now support cancellation via context.
- **Documentation**
  - Updated CHANGELOG to reflect new cancellation feature for snapshot rewriting.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## UNRELEASED
+
+- [#]() memiavl cancel background snapshot rewriting when graceful shutdown.
```
Comment (yihuang) [inline CHANGELOG.md:5]:
```suggestion
- [#1292](https://github.com/crypto-org-chain/cronos/pull/1292) memiavl cancel background snapshot rewriting when graceful shutdown.
```

File: memiavl/db.go [hunk 2]
```diff
@@ -813,11 +837,16 @@ func (db *DB) UpdateCommitInfo() {
 }
 
 // WriteSnapshot wraps MultiTree.WriteSnapshot to add a lock.
-func (db *DB) WriteSnapshot(dir string) error {
+func (db *DB) WriteSnapshot(ctx context.Context, dir string) error {
+	return db.WriteSnapshotWithContext(ctx, dir)
```
Comment (yihuang) [inline memiavl/db.go:841]:
```suggestion
func (db *DB) WriteSnapshot(dir string) error {
	return db.WriteSnapshotWithContext(context.Background(), dir)
```

**Review by mmsqe — APPROVED** [review]
LGTM, is it the fix for https://github.com/crypto-org-chain/cronos/pull/1180 before?

**Comment by yihuang** [comment]
> LGTM, is it the fix for #1180 before?

maybe

File: memiavl/snapshot.go [hunk 3]
```diff
@@ -24,6 +25,9 @@ const (
 	FileNameLeaves   = "leaves"
 	FileNameKVs      = "kvs"
 	FileNameMetadata = "metadata"
+
+	// check for cancel every 1000 leaves
+	CancelCheckInterval = 1000
```
Comment (yihuang) [inline memiavl/snapshot.go:30]:
```suggestion
	// check for cancel every 10000 leaves
	CancelCheckInterval = 10000
```

Comment (yihuang) [reply memiavl/snapshot.go:30]: checking the performance impact of different numbers.

Comment (yihuang) [reply memiavl/snapshot.go:30]: no difference at all.

**Comment by yzang2019** [comment]
Thanks for fixing this!

**Comment by mmsqe** [comment]
> > LGTM, is it the fix for #1180 before?
> 
> maybe

Yes, can't reproduce with smaller IdleTimeout now

**Comment by yzang2019** [comment]
Confirmed this fixed the panic during shutdown

**Review by yzang2019 — APPROVED** [review]

**Pull request merged**